### DISCUSSION TITLE
Generate JSON schema from JSON data

### DISCUSSION THREAD

**Discussion by AlexRMU** [josdejong/svelte-jsoneditor#282] [Ideas]
- you can try to create a schema for a document and, if it turned out, offer the user to apply it
- the scheme specified by the user or analyzed can be automatically saved and used in auto-completion, table columns, etc.
- when you change the schema, the document will be automatically updated to match it and vice versa

https://stackoverflow.com/questions/7341537/tool-to-generate-json-schema-from-json-data
https://github.com/mongodb-js/mongodb-schema

**Comment by josdejong** [comment]
Generating a JSON Schema from a JSON document can be a useful feature. I'm in doubt whether this is common enough to build it in the editor by default though. 

You can use a ready-made library for this and add a button for this in `svelte-jsoneditor` with a few lines of code using `onRenderMenu`.

It may be interesting to add such a button in the web application https://jsoneditoronline.org, what do you think?

**Reply by AlexRMU** [reply]
I think that schemas, even if few people use them, are one of the necessary functions of the json editor, as well as formatting.

**Reply by josdejong** [reply]
JSON Schema is very useful indeed. My experience of maintaining a JSON Editor for 13+ years is that formatting is the most used feature, and JSON Schema isn't used broadly. Generating a JSON Schema from a document is even more of a niche. This is probably a nice candidate for [a plugin](https://github.com/josdejong/svelte-jsoneditor/discussions/284) 😉 .

**Reply by josdejong** [reply]
This feature may be a good candidate to implement in https://jsoneditoronline.org too (reminder to myself 😉).

**Comment by AlexRMU** [comment]
Pay attention to
https://datatracker.ietf.org/doc/rfc8927/
https://jsontypedef.com/
https://github.com/glideapps/quicktype
